exporter/metrics.go

```diff
@@ -2,30 +2,114 @@ package exporter
+	//   mongodb_ss_opcounters{legacy_op_type="update"} 14
+	//   mongodb_ss_opcounters{legacy_op_type="delete"} 22
+	//   mongodb_ss_opcounters{legacy_op_type="getmore"} 9141
+	//   mongodb_ss_opcounters{legacy_op_type="command"} 67923
+	//
+	nodeToPDMetrics = map[string]string{ //nolint
+		"collStats.storageStats.indexDetails.":            "index_name",
+		"globalLock.activeQueue.":                         "count_type",
+		"globalLock.locks.":                               "lock_type",
+		"serverStatus.asserts.":                           "assert_type",
+		"serverStatus.connections.":                       "conn_type",
+		"serverStatus.globalLock.currentQueue.":           "count_type",
+		"serverStatus.metrics.commands.":                  "cmd_name",
+		"serverStatus.metrics.cursor.open.":               "csr_type",
+		"serverStatus.metrics.document.":                  "doc_op_type",
+		"serverStatus.opLatencies.":                       "op_type",
+		"serverStatus.opReadConcernCounters.":             "concern_type",
+		"serverStatus.opcounters.":                        "legacy_op_type",
+		"serverStatus.opcountersRepl.":                    "legacy_op_type",
+		"serverStatus.transactions.commitTypes.":          "commit_type",
+		"serverStatus.wiredTiger.concurrentTransactions.": "txn_rw_type",
+		"serverStatus.wiredTiger.perf.":                   "perf_bucket",
+		"systemMetrics.disks.":                            "device_name",
+	}
+
+	// Regular expressions used to make the metric name Prometheus-compatible
+	// This variables are global to compile the regexps only once.
+	specialCharsRe      = regexp.MustCompile(`[^a-zA-Z0-9_]+`) //nolint
+	repeatedUnderscores = regexp.MustCompile(`__+`)            //nolint
+	dollarRe            = regexp.MustCompile(`\_$`)            //nolint
+)
+
+func prometheusize(s string) string {
+	// Rules to shrink metric names
+	// Please do not change the definitions order: rules are sorted by precedence.
+	prefixes := [][]string{
+		{"serverStatus.wiredTiger.transaction", "ss_wt_txn"},
+		{"serverStatus.wiredTiger", "ss_wt"},
+		{"serverStatus", "ss"},
+		{"replSetGetStatus", "rs"},
+		{"systemMetrics", "sys"},
+		{"local.oplog.rs.stats.wiredTiger", "oplog_stats_wt"},
+		{"local.oplog.rs.stats", "oplog_stats"},
+		{"collstats_storage.wiredTiger", "collstats_storage_wt"},
+		{"collstats_storage.indexDetails", "collstats_storage_idx"},
+		{"collStats.storageStats", "collstats_storage"},
+		{"collStats.latencyStats", "collstats_latency"},
+	}
+
+	for _, pair := range prefixes {
+		if strings.HasPrefix(s, pair[0]+".") {
+			s = pair[1] + strings.TrimPrefix(s, pair[0])
+			break
 		}
 	}
-	return b.String()
+
+	s = specialCharsRe.ReplaceAllString(s, "_")
+	s = dollarRe.ReplaceAllString(s, "")
+	s = repeatedUnderscores.ReplaceAllString(s, "_")
+
+	return exporterPrefix + s
 }
 
-func makeRawMetric(name string, value interface{}) (prometheus.Metric, error) {
+func rename(prefix, name string) (string, string) {
```

Comment:
this function doesn't look like only renaming metrics, I would say it's like getting label. So looks like it's better to extract it from here to a place where it's called